pkg/controller/master-controller-manager/external-cluster/cluster_controller.go

```diff
@@ -0,0 +1,138 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package externalcluster
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+
+	kubermaticapiv1 "github.com/kubermatic/kubermatic/pkg/api/v1"
+	kubermaticv1 "github.com/kubermatic/kubermatic/pkg/crd/kubermatic/v1"
+	kuberneteshelper "github.com/kubermatic/kubermatic/pkg/kubernetes"
+	"github.com/kubermatic/kubermatic/pkg/resources"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ControllerName = "external_cluster_controller"
+)
+
+// Reconciler is a controller which is responsible for managing clusters
+type Reconciler struct {
+	ctx context.Context
+	ctrlruntimeclient.Client
+	log *zap.SugaredLogger
+}
+
+// Add creates a cluster controller.
+func Add(
+	ctx context.Context,
+	mgr manager.Manager,
+	log *zap.SugaredLogger) error {
+	reconciler := &Reconciler{
+		log:    log.Named(ControllerName),
+		Client: mgr.GetClient(),
+		ctx:    ctx,
+	}
+	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return err
+	}
+	// Watch for changes to ExternalCluster
+	err = c.Watch(&source.Kind{Type: &kubermaticv1.ExternalCluster{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+	return nil
+
+}
+
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	resourceName := request.Name
+	log := r.log.With("request", request)
+	log.Debug("Processing")
+
+	icl := &kubermaticv1.ExternalCluster{}
+	if err := r.Get(r.ctx, client.ObjectKey{Namespace: metav1.NamespaceAll, Name: resourceName}, icl); err != nil {
+		if kerrors.IsNotFound(err) {
+			log.Debug("Could not find imported cluster")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if icl.DeletionTimestamp != nil {
+		if kuberneteshelper.HasOnlyFinalizer(icl, kubermaticapiv1.ExternalClusterKubeconfigCleanupFinalizer) {
+			if err := r.cleanUpKubeconfigSecret(icl); err != nil {
+				log.Errorf("Could not delete kubeconfig secret, %v", err)
+				return reconcile.Result{}, err
+			}
+		}
+	}
```

Comment:
If this is the only logic, this should be achievable by just using an owner reference, no?

Reply:
This is a similar situation like with Cluster and Credentials Secret. In create method we first create a secret and then cluster, so we don't have a reference because the cluster doesn't exist yet. We can't store sensitive data like credentials or kubeconfig in cluster or externalCluster object event for a second. First, we store the sensitive data into secret then we create cluster. This is how it works right now.

Reply:
I appreciate that, but the order of creation does not mean that in the future, you couldn't just create the secret, create the cluster, then update the secret to add the owner reference.

Reply:
1 endpoint creates a secret with a special label (atomic operation)
2 external_cluster_controller watches secrets with this label
3 reconcile method creates an external cluster with owner reference 

sounds reasonable?

Reply:
Sounds good to me.